### PR TITLE
core: make idle thread optional

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR makes the idle thread optional, if the architecture implements support for that.
In order to do that, an architecture needs to implement `sched_arch_idle()` and disable the module `core_idle_thread`.

The main benefits are:

- saving a full thread context and stack
- faster and shorter sleep transitions (no need to schedule idle thread first)

The PR includes an implementation for CortexM3 and higher (M0+ doesn't support dynamically changing IRQ priorities, which the current implementation relies on).

### Testing procedure

Any threading application obviously still needs to work. Also, the pm machinery needs to work as before.

Any shell test should confirm that the scheduler is idling (not switching to any thread if there's no runnable) and reacting to ISR (react to uart input and re-schedule shell thread).

I tested power management using tests/periph_pm, confirming that unblocking the right mode for a while makes the shell unresponsive (power mode has been set), and that when the RTC interrupt triggers, shell works again.

*edit* the commit history looks a bit scary due to #13825 being included. The new logic of this PR is contained in the latest two commits (one for basic support in core, one for the cortexm implementation).

### Issues/PRs references

This PR needs #13825, which re-writes the Cortex-M PendSV logic so it can handle the case where there's no thread to save on entry.

### CODE REVIEW

**Comment by bergzand** [comment]
This breaks the schedstatistics module :cry:

**Comment by maribu** [comment]
> disable the module `core_idle_thread`.

I would like to inverse the logic. Could we instead have a feature (and corresponding module) such as `thread_less_idling`, which would always be in `FEATURS_OPTIONAL`?

Providing a feature instead of disabling a module seems more natural to me.

**Comment by kaspar030** [comment]
> Providing a feature instead of disabling a module seems more natural to me.

done

**Comment by kaspar030** [comment]
> This breaks the schedstatistics module cry

This'll need reworking the sched_cb logic: `sched_cb(sched_active_pid, next_thread->pid);` expects ending of one and starting the next thread to happen at the same time.

File: core/sched.c [hunk 1]
```diff
@@ -80,11 +80,14 @@ int __attribute__((used)) sched_run(void)
 {
     sched_context_switch_request = 0;
 
+#ifndef MODULE_CORE_IDLE_THREAD
+    while (!runqueue_bitcache) {
```
Comment (kaspar030) [inline core/sched.c:84]: note to self: as is, the compiler must assume that sched_arch_idle() calls a function that modifies runqueue_bitcache (it is static in this compilation unit). With LTO, it might come to a different result. need to make this atomic or volatile.

**Comment by kaspar030** [comment]
I marked this WIP, as there's some code that expects that there's always an active thread (e.g., most of the IPC code due to sched_switch()).

**Comment by fjmolinas** [comment]
this one needs a rebase now :)

**Comment by kaspar030** [comment]
> I marked this WIP, as there's some code that expects that there's always an active thread (e.g., most of the IPC code due to sched_switch()).

It seems to me that all that code is actually called when within thread context, meaning that having no active thread won't happen. So I've removed WIP.

**Comment by kaspar030** [comment]
@leandrolanzieri I need some suggestions on where to put the new "no_idle_thread" feature.

**Comment by leandrolanzieri** [comment]
> @leandrolanzieri I need some suggestions on where to put the new "no_idle_thread" feature.

I think that the definition of `HAS_NO_IDLE_THREAD` can be placed in `/kconfigs/Kconfig.features`.

File: cpu/cortexm_common/Makefile.features [hunk 2]
```diff
@@ -30,3 +30,8 @@ else ifeq ($(CPU_CORE),cortex-m23)
 else
   $(error Unkwnown cortexm core: $(CPU_CORE))
 endif
+
+# cortex-m3 and higher don't need the idle thread
+ifneq (,$(filter cortex-m2% cortex-m4% cortex-m3% cortex-m7%,$(CPU_ARCH)))
```
Comment (leandrolanzieri) [inline cpu/cortexm_common/Makefile.features:35]:
```suggestion
ifneq (,$(filter cortex-m2% cortex-m4% cortex-m3% cortex-m7%,$(CPU_CORE)))
```

Comment (leandrolanzieri) [reply cpu/cortexm_common/Makefile.features]: On second thoughts, maybe checking for the architecture is cleaner?
```Makefile
ifneq (,$(filter armv7 armv8,$(CPU_ARCH)))
```

Comment (kaspar030) [reply cpu/cortexm_common/Makefile.features]: I don't mind, personally I'm thinking "cortex-m3 and higher" instead of "armv7". your call.

Comment (leandrolanzieri) [reply cpu/cortexm_common/Makefile.features]: I would go with the arch approach just to be in line with what you added in the cortexm_common Kconfig.

Comment (kaspar030) [reply cpu/cortexm_common/Makefile.features]: ok

Comment (leandrolanzieri) [reply cpu/cortexm_common/Makefile.features]: Oh I think the correct values are `armv7m` and `armv8m` :/

Comment (kaspar030) [reply cpu/cortexm_common/Makefile.features]: :) fixed

**Comment by kaspar030** [comment]
We had some offline discussions on this PR. I checked the core code, this shouldn't(tm) cause issues. @fjmolinas is running a full test run on iotlab-m3. I think, should that not cause any issues, we should merge this, so it get's some use over the next weeks before release.

Should this cause any non-fixable issues (or any trouble during release testing), this can be disabled by removing the "FEATURES_OPTIONAL" that pulls it in in the first place.

@miri64 does that sound sensible to you?

Just to assess the impact, some parts of the code are slightly changed for all platforms (sched_cb handling KERNEL_PID_UNDEF, and factoring out `_unschedule()` in `sched_run()`).
The jucy bits are only enabled on Cortex-M3 and higher.

**Comment by miri64** [comment]
> Should this cause any non-fixable issues (or any trouble during release testing), this can be disabled by removing the "FEATURES_OPTIONAL" that pulls it in in the first place.
> 
> @miri64 does that sound sensible to you?

Yes. I would however still insist on not merging this after soft feature freeze.

**Comment by kaspar030** [comment]
> Yes. I would however still insist on not merging this after soft feature freeze.

Agreed, this is under the assumption that it gets ACK^2'ed this week.

**Comment by fjmolinas** [comment]
I launched the tests in  `04-single-hop-6lowpan-icmp` that target `iotlab-m3` boards so 1-4 and 9-10. All tests where passing.

**Comment by kaspar030** [comment]
I pushed a fix for tests/shell.
(and I accidentally rebased while doing so)

File: sys/schedstatistics/schedstatistics.c [hunk 3]
```diff
@@ -31,15 +31,18 @@ void sched_statistics_cb(kernel_pid_t active_thread, kernel_pid_t next_thread)
 {
     uint32_t now = xtimer_now().ticks32;
 
-    /* Update active thread runtime, there is always an active thread since
-       first sched_run happens when main_trampoline gets scheduled */
-    schedstat_t *active_stat = &sched_pidlist[active_thread];
-    active_stat->runtime_ticks += now - active_stat->laststart;
+    /* Update active thread stats */
+    if (active_thread != KERNEL_PID_UNDEF) {
```
Comment (bergzand) [inline sys/schedstatistics/schedstatistics.c:35]: Should it also be documented in the SCHED_CB docs that `KERNEL_PID_UNDEF` is possible now?

File: core/init.c [hunk 4]
```diff
@@ -64,17 +67,19 @@ static void *idle_thread(void *arg)
...
 
-static char main_stack[THREAD_STACKSIZE_MAIN];
 static char idle_stack[THREAD_STACKSIZE_IDLE];
+#endif
 
 void kernel_init(void)
 {
     irq_disable();
 
+#ifdef MODULE_CORE_IDLE_THREAD
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply core/init.c:77]: sched_arch_idle() would be undefined in that case?

File: sys/schedstatistics/schedstatistics.c (hunk 3 above)
Comment (kaspar030) [reply sys/schedstatistics/schedstatistics.c:35]: there's not much documentation for sched_cb I think, other than that the whole module does what it does. where would you like to see extra documentation?

File: core/init.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (bergzand) [reply core/init.c:77]: But that doesn't matter for this file right?

File: sys/schedstatistics/schedstatistics.c (hunk 3 above)
Comment (bergzand) [reply sys/schedstatistics/schedstatistics.c:35]: Maybe leave this to a follow-up, there is indeed not much documentation, better do this right in a new PR, I can open something later this week I think

Comment (bergzand) [reply sys/schedstatistics/schedstatistics.c:35]: See #14334

File: core/init.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (miri64) [reply core/init.c:77]: `IS_USED` ;-)

File: core/sched.c [hunk 5]
```diff
@@ -76,15 +76,45 @@ static void (*sched_cb) (kernel_pid_t active_thread,
...
-
     thread_t *active_thread = (thread_t *)sched_active_thread;
 
-    /* The bitmask in runqueue_bitcache is never empty,
-     * since the threading should not be started before at least the idle thread was started.
-     */
+#ifndef MODULE_CORE_IDLE_THREAD
+    if (!runqueue_bitcache) {
+        if (active_thread) {
+            // unschedule
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply core/sched.c:108]: nice one, the function call is missing there...

[the remark this answers is not in the document]
Comment (kaspar030) [reply core/sched.c]: I guess we have to re-run the tests.

[the remark this answers is not in the document]
Comment (fjmolinas) [reply core/sched.c]: Will do

[the remark this answers is not in the document]
Comment (fjmolinas) [reply core/sched.c]: Re-tested, all is still good.

File: core/init.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaspar030) [reply core/init.c]: changed to IS_USED, here and in sched.c

**Comment by kaspar030** [comment]
- rebased to include the sched_cb doc change (which is actually inaccurate without this PR)
- changed to use IS_USED instead of ifdefs. did some comparison on samr21-xpro (which does use the idle thread), no size diff, so I'm assuming the linker does its job.

File: tests/shell/tests/01-run.py [hunk 6]
```diff
@@ -60,6 +59,8 @@
     ('end_test', '[TEST_END]'),
 )
 
+CMDS_REGEX = { 'ps' }
```
Comment (bergzand) [inline tests/shell/tests/01-run.py:62]: Flake8 wants it like this:
```suggestion
CMDS_REGEX = {'ps'}
```

**Comment by bergzand** [comment]
Please Squash!

File: tests/shell/tests/01-run.py (hunk 6 above)
Comment (bergzand) [reply tests/shell/tests/01-run.py:62]: Feel free to amend this directly

Comment (fjmolinas) [reply tests/shell/tests/01-run.py:62]: I foresee a conflict with #13197, as a headsup.

**Comment by kaspar030** [comment]
squashed

File: tests/shell/tests/01-run.py (hunk 6 above)
Comment (kaspar030) [reply tests/shell/tests/01-run.py]: seems like git manages to merge this just fine. :smile:

Comment (kaspar030) [reply tests/shell/tests/01-run.py]: fixed

File: core/include/sched.h [hunk 7]
```diff
@@ -215,13 +229,6 @@ NORETURN void sched_task_exit(void);
  * @param   next        Pid of the next scheduled thread
  */
 typedef void (*sched_callback_t)(kernel_pid_t active, kernel_pid_t next);
-
-/**
- * @brief  Register a callback that will be called on every scheduler run
- *
- * @param[in] callback The callback functions that will be called
- */
-void sched_register_cb(sched_callback_t callback);
```
Comment (fjmolinas) [inline core/include/sched.h:224]: Was this removal intentional?

Comment (kaspar030) [reply core/include/sched.h]: nope, I put it back and amended

**Comment by fjmolinas** [comment]
I believe failing tests are due to a faulty worker. @kaspar030 anyway that could take a look at FU?

**Comment by fjmolinas** [comment]
These are the failing tests, I'll re-run locally"

<details>

```
    - tests/driver_apds99xx_full
    - tests/cond_order
    - tests/driver_apds99xx
    - tests/evtimer_msg
    - tests/devfs
    - examples/micropython
    - tests/pkg_libb2
    - tests/pkg_tinycbor
    - tests/bench_runtime_coreapis
    - tests/pkg_nanopb
    - tests/progress_bar
    - tests/test_tools
    - tests/pthread_condition_variable
    - tests/pkg_ucglib
    - tests/pkg_yxml
    - tests/rmutex_cpp
    - tests/posix_time
    - tests/trace
    - tests/periph_hwrng
    - tests/trickle
    - tests/ps_schedstatistics
    - tests/rmutex
    - tests/isr_yield_higher
    - tests/pkg_umorse
    - tests/xtimer_usleep
    - tests/gnrc_ndp
    - tests/xtimer_remove
    - tests/xtimer_usleep_short
    - tests/heap_cmd
    - tests/thread_flags
    - tests/shell
    - tests/thread_cooperation
    - tests/thread_race
    - tests/thread_msg_bus
    - tests/pkg_spiffs
    - tests/ssp
    - tests/sched_testing
    - tests/thread_exit
    - tests/thread_msg_block_wo_queue
    - tests/xtimer_hang
    - tests/thread_msg
    - tests/pipe
    - tests/vfs_plus_stdio
    - tests/xtimer_now32_overflow
    - tests/unittests
    - tests/pkg_monocypher
    - tests/pthread_barrier
    - tests/cpp_ext
    - tests/cpp11_thread
    - tests/bloom_bytes
```
</details>

**Comment by kaspar030** [comment]
> I believe failing tests are due to a faulty worker. @kaspar030 anyway that could take a look at FU?

I've restarted the usual suspect, let me re-trigger here.

**Comment by kaspar030** [comment]
Turns out the unittest hard fault was caused by too a too small stack for main. I've added a commit to increase it slightly. (stack is also small on master, depending on compiler, it hard faults there without increase. This was unnoticed as the crash happens after finishing the tests).

**Comment by miri64** [comment]
This PR still needs two ACKs to be merged before soft feature freeze (which I will anounce in a few hours) ;-).

**Comment by bergzand** [comment]
And go!

**Comment by kaspar030** [comment]
> And go!

Thanks for reviewing!

**Comment by miri64** [comment]
Now let's rethink making the [`main` thread optional](https://github.com/RIOT-OS/RIOT/pull/10971) ;-P

**Comment by fjmolinas** [comment]
Bisecting I have 0ff9e554eb712fe4f37ea46e232eaed9c20b9071 causing `tests/xtimer_periodic_wakeup` to get stuck.

```
Testing interval 25... (now=4054509)
Testing interval 24... (now=4057901)
Testing interval 23... (now=4061293)
Testing interval 22... (now=4064686)
Testing interval 21... (now=4068078)
Testing interval 20... (now=4071461)
Testing interval 19... (now=4074844)
Testing interval 18... (now=4078229)
Testing interval 17... (now=4081612)
Testing interval 16... (now=4084996)
# gets stuck here
```

**Comment by bergzand** [comment]
Totally not related to #14557, some measurements on the same54-xpro with bench_mutex_pingpong:

| | idle thread | no_idle_thread | impact|
|-|-|-|-|
| GCC | 262582 | 251572 | 4.1% |
| GCC+LTO | 272109 | 252631 | 7.1% |
|LLVM | 260303 | 244897 | 5.9% |
